File: golang/275_h_index/h_index.go
/*
 * @Date: 2021-07-12 08:06:52
 * @Author: Mengsen Wang
 * @LastEditors: 854284842@qq.com
 * @LastEditTime: 2023-10-30
 */

// Package main ...
package main

import (
	"sort"
	"testing"

	"github.com/stretchr/testify/assert"
)

func hIndex(citations []int) int {
	n := len(citations)
	return n - sort.Search(n, func(x int) bool { return citations[x] >= n-x })
}

func main() {
	tests := []struct {
		citations []int
		ans       int
	}{
		{[]int{0, 1, 3, 5, 6}, 3},
		{[]int{1, 2, 100}, 2},
	}

	for index, test := range tests {
		assert.Equal(&testing.T{}, test.ans, hIndex(test.citations), index)
	}
}

File: rust/274_h_index/h_index.rs
/*
 * @Date: 2021-07-12 08:24:46
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2021-07-12 08:26:22
 */

fn h_index(mut citations: Vec<i32>) -> i32 {
    citations.sort();
    let len = citations.len();
    for x in citations.iter().enumerate().map(|(i, value)| {
        if *value as usize >= len - i {
            return len - i;
        }
        return 99999999;
    }) {
        if x != 99999999 {
            return x as i32;
        }
    }
    return 0;
}

fn main() {
    let citations = vec![0, 1, 3, 5, 6];
    assert_eq!(h_index(citations), 3);

    let citations = vec![1];
    assert_eq!(h_index(citations), 1);
}
